Gems/Terrain/Code/Source/TerrainRenderer/Components/TerrainSurfaceMaterialsListComponent.cpp

```diff
@@ -327,6 +327,21 @@ namespace Terrain
     }
 
     void TerrainSurfaceMaterialsListComponent::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        // REMARK: This function is typically invoked within the context of one of the AssetManager::OnAssetXXX functions,
+        // which may hold a mutex, in turn when instantiating StreamingImages, which is a multi-threaded operation,
+        // a deadlock may arise when one of those threads release another StreamingImage because that class executes
+        // Data::AssetBus::MultiHandler::BusDisconnect(GetAssetId()); which will acquire the same mutex of the original
+        // AssetManager::OnAssetXXX and triggers a deadlock.
+        // The solution is to enqueue texture update on the next tick.
+        auto postTickLambda = [=]()
+        {
+            OnAssetReadyPostTick(asset);
+        };
+        AZ::SystemTickBus::QueueFunction(AZStd::move(postTickLambda));
+    }
+
+    void TerrainSurfaceMaterialsListComponent::OnAssetReadyPostTick(AZ::Data::Asset<AZ::Data::AssetData> asset)
```

Comment:
Was this function also doing MultiHandler::BusDisconnect? Can you explain why are we seeing deadlock related to AssetManager?

Reply:
As explained in the commit message:
```
The second most important change pertains to some components that instantiate StreamingImage(s) objects in the same thread when the AssetManager calls one of the OnAssetReady() or OnAssetReloaded() functions. This can also cause deadlocks, because when AsyncUploadQueue() is called, and an StreamingImage is released during the Secondary Copy Queue thread, the StreamingImage class would call:

Data::AssetBus::MultiHandler::BusDisconnect(GetAssetId());

Which will deadlock when it acquires the AssetManager container mutex
```